File: backend-node/src/routes/jobs.js
import { Router } from 'express';
import { db } from '../db/database.js';

const router = Router();

router.get('/:id', (req, res) => {
  const job = db.prepare('SELECT * FROM generation_jobs WHERE id=?').get(req.params.id);
  if (!job) return res.status(404).json({ detail: 'Job not found' });
  res.json(job);
});

router.get('/:id/stream', (req, res) => {
  const jobId = req.params.id;

  res.setHeader('Content-Type', 'text/event-stream');
  res.setHeader('Cache-Control', 'no-cache');
  res.setHeader('X-Accel-Buffering', 'no');
  res.flushHeaders();

  const send = (data) => res.write(`data: ${JSON.stringify(data)}\n\n`);

  const poll = setInterval(() => {
    const job = db.prepare('SELECT * FROM generation_jobs WHERE id=?').get(jobId);
    if (!job) {
      send({ error: 'Job not found' });
      res.write('event: done\ndata: {}\n\n');
      clearInterval(poll);
      res.end();
      return;
    }
    send({ id: job.id, status: job.status, progress: job.progress_message, error: job.error_message });
    if (job.status === 'COMPLETE' || job.status === 'FAILED') {
      res.write('event: done\ndata: {}\n\n');
      clearInterval(poll);
      res.end();
    }
  }, 1500);

  req.on('close', () => clearInterval(poll));
});

export default router;

File: backend/app/api/routes/jobs.py
"""Job status polling and SSE stream."""
from __future__ import annotations

import asyncio
import json

from fastapi import APIRouter, Depends
from fastapi.responses import StreamingResponse

from app.api.deps import get_job_repo
from app.db.repository import JobRepository
from app.db.session import AsyncSessionLocal
from app.models.models import GenerationJob
from app.schemas.schemas import JobRead

router = APIRouter()


@router.get("/{job_id}", response_model=JobRead)
async def get_job(job_id: str, repo: JobRepository = Depends(get_job_repo)):
    return await repo.get_or_raise(job_id)


@router.get("/{job_id}/stream")
async def stream_job(job_id: str):
    """Server-Sent Events stream — polls DB every 1.5 s until COMPLETE or FAILED."""

    async def generator():
        try:
            while True:
                async with AsyncSessionLocal() as db:
                    job = await db.get(GenerationJob, job_id)

                if job is None:
                    yield f"data: {json.dumps({'error': 'Job not found'})}\n\n"
                    yield "event: done\ndata: {}\n\n"
                    break

                yield f"data: {json.dumps({'id': str(job.id), 'status': job.status, 'progress': job.progress_message, 'error': job.error_message})}\n\n"

                if job.status in ("COMPLETE", "FAILED"):
                    yield "event: done\ndata: {}\n\n"
                    break

                await asyncio.sleep(1.5)
        except asyncio.CancelledError:
            pass  # client disconnected

    return StreamingResponse(
        generator(),
        media_type="text/event-stream",
        headers={"Cache-Control": "no-cache", "X-Accel-Buffering": "no"},
    )
